lib/semantics/tools.cc

```diff
@@ -508,6 +509,169 @@ std::optional<parser::MessageFixedText> WhyNotModifiable(
+  return std::visit(
+      common::visitors{
+          [](const common::Indirection<parser::EventPostStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::EventWaitStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::FormTeamStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::LockStmt> &) { return true; },
+          [](const common::Indirection<parser::StopStmt> &) { return true; },
+          [](const common::Indirection<parser::SyncAllStmt> &) { return true; },
+          [](const common::Indirection<parser::SyncImagesStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::SyncMemoryStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::SyncTeamStmt> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::UnlockStmt> &) { return true; },
+          [](const common::Indirection<parser::AllocateStmt> &stmt) {
+            return IsImageControlStmt(stmt.value());
+          },
+          [](const common::Indirection<parser::DeallocateStmt> &stmt) {
+            return IsImageControlStmt(stmt.value());
+          },
+          [](const common::Indirection<parser::CallStmt> &stmt) {
+            return IsImageControlStmt(stmt.value());
+          },
+          [](const auto &) { return false; },
+      },
+      actionStmt.statement.u);
+}
+
+bool IsImageControlStmt(const parser::ExecutableConstruct &construct) {
+  return std::visit(
+      common::visitors{
+          [](const common::Indirection<parser::ChangeTeamConstruct> &) {
+            return true;
+          },
+          [](const common::Indirection<parser::CriticalConstruct> &) {
+            return true;
+          },
+          [](const parser::Statement<parser::ActionStmt> &actionStmt) {
+            return IsImageControlStmt(actionStmt);
+          },
+          [](const auto &) { return false; },
+      },
+      construct.u);
+}
+
+std::optional<parser::MessageFixedText> GetImageControlStmtCoarrayMsg(
+    const parser::ExecutableConstruct &construct) {
+  if (const parser::Statement<parser::ActionStmt> *actionStmt{
+          std::get_if<parser::Statement<parser::ActionStmt>>(&construct.u)}) {
+    if (std::holds_alternative<common::Indirection<parser::AllocateStmt>>(
+            actionStmt->statement.u)) {
+      return "ALLOCATE of a coarray is an image control statement"_en_US;
+    } else if (std::holds_alternative<
+                   common::Indirection<parser::DeallocateStmt>>(
+                   actionStmt->statement.u)) {
+      return "DEALLOCATE of a coarray is an image control statement"_en_US;
+    } else if (std::holds_alternative<common::Indirection<parser::CallStmt>>(
+                   actionStmt->statement.u)) {
+      return "MOVE_ALLOC of a coarray is an image control statement"_en_US;
+    }
```

Comment:
`std::visit` is clearer than multiple calls to `std::holds_alternative`.

Reply:
Thanks, Tim.  I'll make this change.